### ISSUE TITLE
Improve metrics and access logs

### ISSUE THREAD

**Issue by FlorentinDUBOIS** [sozu-proxy/sozu#937]
Hello there :wave:,

After some discussions with @StevenLeRoux, @BlackYoup and @Wonshtrum.
We found out that Sōzu need to provides a little more information about request and response.
So, I use this issue to make a summary of what we need.

Firstly, I will remember that we have to tackle the issue #892.

### Access logs

Currently, when we handle request (TCP, HTTP[s] or WS[s]) will write following access logs.

```txt
01H0FXZVQK7RK3W55X7V26M259 TcpTest TcpTest-0-127.0.0.1:4000	127.0.0.1:52378 -> 127.0.0.1:4000	19ms 168μs 79 68522	owner=John, uuid=3f740af1-45fd-45ce-b61f-17bf1a51505f	TCP -
01H0FXX3VTR8T3B1K0ZHXQZZNQ MyCluster the-backend-to-my-app	127.0.0.1:50444 -> 127.0.0.1:1026	61ms 4959μs 79 68535	key=value	HTTPS-TLS1.3 lolcatho.st:8443 200	GET /
```

Below, an explanation of the above access logs:

| request id                   | cluster id  | backend id              | client socket address -> backend socket address | response time | service time | bytes in | bytes out | metadata    | protocol       | hostname           | status | verb  | path |
| ---------------------------- | ----------- | ----------------------- | ----------------------------------------------- | ------------- | ------------ | -------- | --------- | ----------- | -------------- | ------------------ | ------ | ----- | ---- |
| `01H0FXX3VTR8T3B1K0ZHXQZZNQ` | `MyCluster` | `the-backend-to-my-app` | `127.0.0.1:50444 -> 127.0.0.1:1026`             | `61ms`        | `4959μs`     | `79`     | `68535`   | `key=value` | `HTTPS-TLS1.3` | `lolcatho.st:8443` | `200`  | `GET` | `/`  |

| request id                   | cluster id | backend id                 | client socket address -> backend socket address | response time | service time | bytes in | bytes out | metadata                                                | protocol | websocket context |
| ---------------------------- | ---------- | -------------------------- | ----------------------------------------------- | ------------- | ------------ | -------- | --------- | ------------------------------------------------------- | -------- | ----------------- |
| `01H0FXZVQK7RK3W55X7V26M259` | `TcpTest`  | `TcpTest-0-127.0.0.1:4000` | `127.0.0.1:52378 -> 127.0.0.1:4000`             | `19ms`        | `168μs`      | `79`     | `68522`   | `owner=John, uuid=3f740af1-45fd-45ce-b61f-17bf1a51505f` | `TCP`    | `-`               |

> `response time` : the total duration of the request 
> `service time` : the total duration that Sozu took to treat the request
> `websocket context`: the value is `-` if the protocol is `TCP`

You can find the code source that output those logs here:

#### TCP or WS

https://github.com/sozu-proxy/sozu/blob/3aea15ccff60bd8ba680b19b602ac9114fb2370e/lib/src/protocol/pipe.rs#L275-L287
https://github.com/sozu-proxy/sozu/blob/3aea15ccff60bd8ba680b19b602ac9114fb2370e/lib/src/protocol/pipe.rs#L328-L340

#### HTTP  

https://github.com/sozu-proxy/sozu/blob/3aea15ccff60bd8ba680b19b602ac9114fb2370e/lib/src/protocol/http/mod.rs#L614-L628 
https://github.com/sozu-proxy/sozu/blob/3aea15ccff60bd8ba680b19b602ac9114fb2370e/lib/src/protocol/http/mod.rs#L682-L696 
https://github.com/sozu-proxy/sozu/blob/3aea15ccff60bd8ba680b19b602ac9114fb2370e/lib/src/protocol/http/mod.rs#L758-L773

### Metrics

Today, we have a bunch of metrics and we do not use all of them. 
We should investigate. 
To do so, we have to: 
- [ ] make a proper list of metrics (classname+labels+kind+description).
- [ ] Then, establish the cardinality of each one and segregate them in two parts:
  - [ ] The one that are Sōzu-related (event loop, worker, etc..)
  - [ ] The one that are cluster-related (status code, response time, service time, cluster response time, data transfer, etc..)
    - [ ] The cluster-related have to be labeled with the cluster identifier and optionaly the backend identifier.

### Access logs

To us, it miss some informations in access logs as :

- client rtt: the time in millisecond to establish a connection between the client and Sōzu
- server rtt: the time in millisecond to establish a connection between Sōzu and the backend
- cluster response time: the time in millisecond between the end of the transfer of the request body and the response header betweend Sōzu and the backend
- data transfer time: the total duration in millisecond of transfering data between the client to the backend and vice-versa

### Metrics

We should investigate first the item above.

Registry:
- Reference around logs: https://www.haproxy.com/blog/introduction-to-haproxy-logging/
- TCP log format: https://www.haproxy.com/documentation/hapee/1-8r1/onepage/#8.2.2
- HTTP log format: https://www.haproxy.com/documentation/hapee/1-8r1/onepage/#8.2.3
- What is round trip time: https://www.cloudflare.com/learning/cdn/glossary/round-trip-time-rtt/

**Comment by Wonshtrum** [comment]
The first draft can be found on the latest commit of #904.

**Comment by Keksoj** [comment]
Oh and we should do something about this:

```
Worker 0
=========
┌──────────────────────┬───────┬───────┐
│                      │ gauge │ count │
├──────────────────────┼───────┼───────┤
│ accept_queue.count   │ 0     │       │
├──────────────────────┼───────┼───────┤
│ buffer.count         │ 0     │       │
├──────────────────────┼───────┼───────┤
│ bytes_in             │       │ 150   │
├──────────────────────┼───────┼───────┤
│ bytes_out            │       │ 178   │
├──────────────────────┼───────┼───────┤
│ client.connections   │ 0     │       │
├──────────────────────┼───────┼───────┤
│ http.404.errors      │       │ 2     │
├──────────────────────┼───────┼───────┤
│ http.active_requests │ 0     │       │
├──────────────────────┼───────┼───────┤
│ http.requests        │       │ 2     │
├──────────────────────┼───────┼───────┤
│ protocol.http        │ 0     │       │
├──────────────────────┼───────┼───────┤
│ sent_access_logs     │       │ 2     │
├──────────────────────┼───────┼───────┤
│ slab.count           │ 6     │       │
└──────────────────────┴───────┴───────┘
```

Notice that many `.count` are actually gauges. This is *wrong*.

**Comment by FlorentinDUBOIS** [comment]
## Status update — 2026-05-02 (HEAD = `main` @ `8c883bb5`)

PR #1209 (`feat/h2-mux`) merged into `main` on 2026-04-27, and the access-log + metrics commits referenced in my 2026-04-24 update above are now in `main`. The companion metrics-only PR #1196 (`feat/metrics-gaps`) was closed unmerged on 2026-04-28 because its scope had already been folded into the H2-mux series — its sub-issues #597, #1146 and #641 were closed on 2026-05-02. The original umbrella here can now be narrowed considerably; tracking what's left below.

### Already shipped on `main`

Access-log fields (`command/src/command.proto:1456+`, `ProtobufAccessLog` tags 9–29):

| Field | H1 (kawa_h1) | H2 / mux | Pipe (TCP/WS) | TCP frontend |
|---|---|---|---|---|
| `client_rtt` (tag 9) | populated `kawa_h1/mod.rs:1045` | **None — TODO at `mux/stream.rs:289`** | populated `pipe.rs:322` | populated `tcp.rs:282` |
| `server_rtt` (tag 10) | populated `kawa_h1/mod.rs:1046` | **None — TODO at `mux/stream.rs:290`** | populated `pipe.rs:323` | None |
| `user_agent` (tag 16) | populated | populated | None | None |
| `x_request_id` (tag 24) | populated | populated | n/a | n/a |
| `tls_version`/`cipher`/`sni`/`alpn` (tags 25–28) | populated | populated | n/a | n/a |
| `xff_chain` (tag 29) | populated | populated | n/a | n/a |
| `traceparent` (otel feature) | populated | populated | n/a | n/a |

Originating commits: `ed1966e1` (x-request-id), `2e7d9434` (TLS+XFF), `db7c33ec` (OTel propagation), `f76c5d5b` (configurable correlation header — closes #1145).

Metrics:

- `12be94c0` per-status-code HTTP counters across H1+H2 (closed #1146).
- `36d9cffb` event-loop buffer / connection / slab gauge rename + new gauges (closed #597).
- `13f8f296` `http.errors` cluster + backend label.
- `deeba79c` restored mux backend metric parity with H1.
- `a650ad69`, `5f884551` gauge underflow + service-time inflation fixes (closed #641).
- `868fc227` comprehensive metrics reference, 137 metrics catalogued in `doc/configure.md`.

Cluster + backend metric labelling is broader than the original umbrella suggested. The infrastructure (`MetricDetailLevel::{Process,Frontend,Cluster,Backend}` at `command/src/config.rs:1116-1120`, filter at `lib/src/metrics/mod.rs:43`) is in place and is consumed by the `record_backend_metrics!` macro at `lib/src/metrics/mod.rs:474-490` (covers `bytes_in`, `bytes_out`, `backend_response_time`, `backend_connection_time`, `requests`), the `http.errors` and `http.status.*` counters, the default-answer counters in `kawa_h1/mod.rs:1104-1165` and `mux/answers.rs:224-239`, plus the `connections_per_backend` / `backend.up` / `backend.down` / `backend.connections.error` / `backend.connect.retries_exhausted` family across `tcp.rs:559-641,914-943`, `kawa_h1/mod.rs:1310-1333,1664-1742`, `mux/router.rs:119-123,479-518`, `mux/mod.rs:815-819,931-947,1704-1708`, `mux/connection.rs:380-385`. `request_time`/`service_time` (`lib.rs:1273-1275`), `frontend_matching_time` (`http.rs:665-666`, `https.rs:860-861`) and `access_logs.count` (`lib.rs:1293`) are all cluster-labelled.

### Outstanding under this umbrella

1. **H2/mux access-log `client_rtt` / `server_rtt`.** `lib/src/protocol/mux/stream.rs:289-290` has the call wired but commented out (`f27a449d`). Mux `Stream` has no direct `backend_socket` field; the socket is reachable via `Stream::state == StreamState::Linked(token)` → `Context.backend_streams` (`mux/mod.rs:342-350,471-486`) → `Router.backends: HashMap<Token, Connection<SessionTcpStream>>` (`mux/router.rs:70-72`) → `Connection::{H1,H2}.socket` (`mux/connection.rs:54-57`, `mux/h1.rs:108-115`, `mux/h2.rs:1251-1269`). Deciding between per-stream snapshot and per-connection RTT, and plumbing that into `Stream::generate_access_log`, is the residual concrete work. This is the single largest visible asymmetry between H1 and H2 access logs today.

2. **`data_transfer_time`.** Not in the schema. Current proxy is `request_time - service_time`. Needs a decision: add an explicit field (and a populator that subtracts the two), or document the derivation and leave it out.

3. **`cluster_response_time`.** Maps semantically to `backend_response_time`, already emitted as a metric and as the `response_time` access-log field (tag 21) populated from `metrics.backend_response_time()` at `mux/stream.rs:292` and `kawa_h1/mod.rs:1048`. If we want HAProxy-parity wording, that's a doc/alias decision in `doc/configure.md`, not new instrumentation.

4. **Cardinality classification table.** The 137-metric reference at `doc/configure.md` lists metrics but doesn't yet present them as the explicit "Sōzu-related vs cluster-related" cardinality table the original issue called for. Doc work, not code.

5. **Per-cluster/backend label residual sweep.** Most cluster-scoped metrics now carry labels (see infrastructure paragraph above). What remains is an audit pass to identify any cluster-scoped metric still emitted unlabelled. Likely a short list, not a wholesale conversion.

### Linked issues — re-checked against `main`

- **#892 (backend up/down log).** Largely shipped: `lib/src/health_check.rs:461-492` logs UP/DOWN transitions and emits `health_check.up` / `health_check.down`; server wires it at `server.rs:889-890,1448-1457,1547`. Retry-policy paths log transitions independently at `tcp.rs:580-641`, `kawa_h1/mod.rs:1683-1742`, `mux/mod.rs:807-819,929-947`. Probably closeable — proposing to close as completed once @StevenLeRoux / @BlackYoup confirm the existing log lines satisfy what they need.
- **#804 (regex / lookup-time metrics).** Partially shipped: `regex_matching_time` is emitted for domain and path regexes at `lib/src/router/mod.rs:580-584,660-664`; `frontend_matching_time` at `http.rs:665-666` and `https.rs:860-861`. Only a dedicated `router.lookup_time` / non-regex matching-time metric is missing.

### Proposed next steps

I'd like to close out this umbrella in two passes:

1. A small follow-up MR for the H2/mux RTT lift (item 1 above) — single-purpose, matches H1/Pipe behavior, no schema change.
2. A doc-only pass that maps `response_time` to the HAProxy `cluster_response_time` notion, presents the cardinality table (item 4), and decides on `data_transfer_time` either way (item 2).

Once those land, the residual is just (5) — an audit pass — and the linked issues #892/#804 can be re-scoped to their respective small remnants. Happy to keep this issue open as the meta-tracker until those land, or split off a successor — whichever you prefer.

**Comment by FlorentinDUBOIS** [comment]
Closing — covered by #1245 (per-cluster availability + access-log timeout discriminator), #1242 (H2 client/server RTT on access log), #1225 (H2 flood/frame/error counters + TLS handshake + access-log enrichment), and #1233 (metrics gaps rewrite). Reopen for a specific gap still open.
